spring-integration-file/src/main/java/org/springframework/integration/file/splitter/FileSplitter.java

```diff
@@ -224,6 +226,23 @@ public Object next() {
 		}
 	}
 
+	@Override
+	protected void addHeaders(Message<?> message, AbstractIntegrationMessageBuilder<?> builder) {
+		if (message.getPayload() instanceof File) {
+			File file = (File) message.getPayload();
+			builder
+				.setHeaderIfAbsent(FileHeaders.ORIGINAL_FILE, file)
+				.setHeaderIfAbsent(FileHeaders.FILENAME, file.getName());
+
+		}
+		else if (message.getPayload() instanceof String) {
+			File file = new File((String) message.getPayload());
```

Comment:
Yes, makes sense as well. 
That's great that we can have a choice for the target solution, unlike many other tasks without no one idea :smile: 

OK. I agree with you and we really can leave `AbstractIntegrationMessageBuilder` weakness to the end-user, because it won't be already our code. More support for silly SO questions, but better performance on XD.

So, let's change this algorithm only to that `headersCallback` to get headers to populate only once, not on the each iteration!